### DISCUSSION TITLE
Why is OnMessage not planned?

### DISCUSSION THREAD

**Discussion by trajano** [spyoungtech/ahk#319] [Q&A] [answered]
Just curious, because I was thinking of doing something like a "close" window handler specifically for Discord to perform minimize instead,

**Comment by spyoungtech** [comment, marked as answer]
Hmm. Taking a closer look at this, I think I decided that status due to a misconception about its use cases. I previously understood [OnMessage](https://www.autohotkey.com/docs/v1/lib/OnMessage.htm#Examples) to be exclusively for receiving messages of GUI application written with AutoHotkey. Currently, besides the few boxes we already have implemented, there aren't any plans to support AutoHotkey's GUI functionality in the library (that's something I can expound on further, though that's a whole other topic), so I initially set this `OnMessage` feature as not planned under that misconception.

Now that I'm looking closer at this, it's clear to me that `OnMessage` has uses outside of AutoHotkey's GUI functionality. I believe this is something I can put on the roadmap. Thanks for bringing it up!

While I'm thinking through the implementation of the feature, it would be helpful for me if you could explain what use cases you might have for this. That way, I can test those use cases as I develop it.

**Reply by trajano** [reply]
Generally to capture pressing the "close" button on the title bar and make it *minimize* instead.  Some apps like Discord when pressing close, closes the app.  They have an option to make it "minimize to tray" but I actually don't want it to go to "tray" because I already allocated a "spot" for it on the task bar.  Same with a few other apps that I want to always remain on the task bar, even if I press "Alt-F4"

Note the following is ChatGPT answer so I haven't really verified it yet
```
; Intercept the close button click to minimize the window instead of closing it

#Persistent
#SingleInstance force

; Hook into the WM_SYSCOMMAND message (0x0112)
OnMessage(0x0112, "OnSysCommand")

OnSysCommand(wParam, lParam, msg, hwnd) {
    ; Check if the command is SC_CLOSE (0xF060)
    if (wParam = 0xF060) {
        ; Minimize the window instead of closing it
        WinMinimize, ahk_id %hwnd%
        return 0  ; Prevent the default close action
    }
}
```

**Reply by trajano** [reply]
If it were me I'd have something like

```
on_sys_command_close(callback):
  on_sys_command(SC_CLOSE, callback)

on_sys_command(sysCommandId, callback):
  on_message(WM_SYSCOMMAND, (wParam, lParam, msg, hwnd) => {
    if (wParam == sysCommandId) {
      callback()
    }
  })

ahk.on_message(messageId, callback): 
  ...
```
The optimization I would do later would be to disallow `OnMessage` on `WM_SysCommand` and instead have a single OnMessage handler that would process the sys commands rather than spawning a new thread for each sys command ID

Then as a user I'd do something like

```python
never_close_apps = [ "Discord.exe", "DBeaver.exe"]
def close_command_pressed(hwnd: int) -> bool:
    if ahk.win_get_process_name("ahk_id %d" % hwnd) in never_close_apps:
        ahk.win_minimize("ahk_id %d" % hwnd)
        return False
    else:
        return True

ahk.on_sys_command_close(callback=close_command_pressed)
```
